lib/fluent/plugin/buffer.rb

```diff
@@ -172,34 +172,46 @@ def metadata(timekey: nil, tag: nil, variables: nil)
       end
 
       # metadata MUST have consistent object_id for each variation
-      # data MUST be Array of serialized events
+      # data MUST be Array of serialized events, or EventStream
       # metadata_and_data MUST be a hash of { metadata => data }
-      def write(metadata_and_data, bulk: false, enqueue: false)
+      def write(metadata_and_data, format: nil, size: nil, enqueue: false)
         return if metadata_and_data.size < 1
         raise BufferOverflowError, "buffer space has too many data" unless storable?
 
         staged_bytesize = 0
         operated_chunks = []
+        unstaged_chunks = {} # metadata => [chunk, chunk, ...]
+        enqueued_chunks = []
 
         begin
           metadata_and_data.each do |metadata, data|
-            write_once(metadata, data, bulk: bulk) do |chunk, adding_bytesize|
+            write_once(metadata, data, format: format, size: size) do |chunk, adding_bytesize|
               chunk.mon_enter # add lock to prevent to be committed/rollbacked from other threads
               operated_chunks << chunk
+              if chunk.unstaged?
+                unstaged_chunks[metadata] ||= []
+                unstaged_chunks[metadata] << chunk
+              end
+              # this value includes bytes of unstaged chunks, but it will be removed by enqueue_unstaged_chunk
```

Comment:
`enqueue_unstaged_chunk` seems to not touch `@stage_size`.
What does "it will be removed" mean?